src/main/java/com/jnape/palatable/shoki/api/MultiSet.java

```diff
@@ -107,4 +112,137 @@
+     * @return the difference {@link MultiSet}
+     */
+    default MultiSet<A> difference(MultiSet<A> other) {
+        return foldLeft(curried(ms -> into(ms::remove)), this, other);
+    }
+
+    /**
+     * Determine if another {@link MultiSet} is a subset of this {@link MultiSet}. A {@link MultiSet} <code>x</code> is
+     * a subset of <code>y</code> if for every <code>a^i</code> in <code>x</code> there exists <code>a^j</code> in
+     * <code>y</code> where ^ denotes the number of occurrences in the multiset.
+     *
+     * @param other the {@link MultiSet} to test whether it's a subset
+     * @return whether other is a subset
+     */
+    default boolean inclusion(MultiSet<A> other) {
+        return and().foldMap(into(this::contains), other);
+    }
+
+    /**
+     * The union of two {@link MultiSet MultiSets}
+     * <code>xs</code> and <code>ys</code> is the {@link MultiSet} of those elements present in both <code>xs</code>
+     * and <code>ys</code> with the occurrences of whichever one had less elements. This is equivalent to
+     * {@link MultiSet#merge} partially applied with {@link com.jnape.palatable.lambda.semigroup.builtin.Min#min}.
+     *
+     * @param other the {@link MultiSet} to intersect with this {@link MultiSet}
+     * @return the union {@link MultiSet}
+     * @see MultiSet#merge
+     */
+    default MultiSet<A> intersection(MultiSet<A> other) {
+        return merge(min(), other);
+    }
+
+
+    /**
+     * The symmetric difference of two {@link MultiSet MultiSets} <code>xs</code> and <code>ys</code> is the
+     * {@link MultiSet} of the occurrence elements present in either <code>xs</code> or <code>ys</code> but not the
+     * other.
+     *
+     * @param other the {@link MultiSet} to symmetrically subtract from this {@link MultiSet}
+     * @return the union {@link MultiSet}
+     * @see MultiSet#merge
+     */
+    default MultiSet<A> symmetricDifference(MultiSet<A> other) {
+        return difference(other).union(other.difference(this));
+    }
+
+    /**
+     * {@link Set} containing the unique elements from the {@link MultiSet}
+     *
+     * @return the unique {@link Set}
+     */
+    Set<Natural, A> unique();
+
+    /**
+     * Combine {@link MultiSet}s with a {@link Semigroup} to handle the combining of the amounts.
+     * Note that if one of the {@link MultiSet}s does not contain an {@link A} it will still pass
+     * {@link com.jnape.palatable.shoki.api.Natural.Zero} to the {@link Semigroup}.
+     *
+     * @param semigroup the {@link Semigroup}
+     * @param other     the other {@link MultiSet}
+     * @return the merged {@link MultiSet}
+     */
+    default MultiSet<A> merge(Semigroup<Natural> semigroup, MultiSet<A> other) {
+        return foldLeft((acc, a) -> semigroup
+                                .apply(get(a), other.get(a)).match(constantly(acc), result -> acc.add(a, result)),
+                        this.difference(this), // Empty but the correct impl.
```

Comment:
Don't @ me.

Reply:
Why do you need to start from empty? Why not just fold `other`, doing a `get` on `this` for the other `a`, applying the values via the `semigroup`? What am I missing here?

Reply:
You need to go over all the keys in both to make sure you still call the semigroup with zero. I could still use this but I'd have to set(via removeall + add) the value with result instead of just add.

Reply:
```
return foldLeft(curried(ms -> into((a, k_) -> {
                    Natural k = ms.get(a);
                    return semigroup.apply(k, k_)
                            .match(__ -> ms.removeAll(a),
                                   nzk -> eq(nzk, k)
                                          ? ms
                                          : lt(nzk, k)
                                            ? k.minus(nzk).flatMap(CoProduct2::projectB)
                                                    .match(constantly(ms), diff -> ms.add(a, diff))
                                            : nzk.minus(k).flatMap(CoProduct2::projectB)
                                                    .match(constantly(ms), diff -> ms.remove(a, diff)));
                })),
                this,
                other);
```

It's painful, but I don't like the idea of the starting `difference`, that seems like a hack.

I'm currently working on making `MultiSet` subtype `Map` at which point you'll get `put(a, NonZero k)` which will make this implementation trivial, so I won't sweat this too much right now.

Reply:
I'll add this in. Just for my own clarity, is it significantly cheaper to put by adding/removing vs removeAll + add.